database/src/ops/block.rs

```diff
@@ -1,89 +1,454 @@
+        macros::doc_error,
+        output::{
+            add_output, add_rct_output, get_rct_num_outputs, remove_output, remove_rct_output,
+        },
+        tx::{add_tx, get_num_tx, remove_tx},
+    },
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        AmountIndex, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, OutputFlags,
+        PreRctOutputId, RctOutput, TxHash,
+    },
+    StorableVec,
+};
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps/adds them to the appropriate database tables.
+///
+#[doc = doc_error!()]
+///
+/// # Panics
+/// This function will panic if:
+/// - `block.height > u32::MAX` (not normally possible)
+/// - `block.height` is not != [`chain_height`]
+///
+/// # Already exists
+/// This function will operate normally even if `block` already
+/// exists, i.e., this function will not return `Err` even if you
+/// call this function infinitely with the same block.
+// no inline, too big.
+pub fn add_block(
+    block: &VerifiedBlockInformation,
+    tables: &mut impl TablesMut,
+) -> Result<(), RuntimeError> {
+    //------------------------------------------------------ Check preconditions first
 
-/// TODO
-pub fn get_block_hash() {
-    todo!()
-}
+    // Cast height to `u32` for storage (handled at top of function).
+    // Panic (should never happen) instead of allowing DB corruption.
+    // <https://github.com/Cuprate/cuprate/pull/102#discussion_r1560020991>
+    let Ok(height) = u32::try_from(block.height) else {
+        panic!("block.height ({}) > u32::MAX", block.height);
+    };
 
-/// TODO
-pub fn get_block_height() {
-    todo!()
-}
+    let chain_height = chain_height(tables.block_heights())?;
+    assert_eq!(
+        block.height, chain_height,
+        "block.height ({}) != chain_height ({})",
+        block.height, chain_height,
+    );
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+    // Expensive checks - debug only.
+    #[cfg(debug_assertions)]
+    {
+        assert_eq!(block.block.serialize(), block.block_blob);
+        assert_eq!(block.block.hash(), block.block_hash);
+        assert_eq!(block.block.txs.len(), block.txs.len());
+        assert_ne!(block.generated_coins, 0);
+        for (i, tx) in block.txs.iter().enumerate() {
+            assert_eq!(tx.tx_blob, tx.tx.serialize());
+            assert_eq!(tx.tx_weight, tx.tx.weight());
+            assert_eq!(tx.tx_hash, tx.tx.hash());
+            assert_eq!(tx.tx_hash, block.block.txs[i]);
+        }
+    }
```

Comment:
Could do `#[cfg(test)]` instead if you don't want to slow down debug builds too much.

Reply:
I would remove the hashes and weight as they both require serialization as well and if they are wrong we are doing something horribly wrong elsewhere also are you sure the reward can never be 0? even with penalties? 

```suggestion
    // Expensive checks - debug only.
    #[cfg(debug_assertions)]
    {
        assert_eq!(block.block.serialize(), block.block_blob);
        assert_eq!(block.block.txs.len(), block.txs.len());
        assert_ne!(block.generated_coins, 0);
        for (i, tx) in block.txs.iter().enumerate() {
            assert_eq!(tx.tx_blob, tx.tx.serialize());
            assert_eq!(tx.tx_hash, block.block.txs[i]);
        }
    }
```